### PR TITLE
[OAP-1528][oap-native-sql] support ColumnarBroadcastJoin

### PR DESCRIPTION
**Pull request by xuechendi**
Signed-off-by: Chendi Xue <chendi.xue@intel.com>

fixes https://github.com/Intel-bigdata/OAP/issues/1528

### CODE REVIEW

**Comment by xuechendi** [comment]
@zhouyuan , will do fallback when AQE is enabled, and tested with all tpch queries in my cluster w/ columnarBHJ enabled. How about we merge this one first, and I can work on AQE support in another PR?

File: oap-native-sql/core/src/main/scala/com/intel/oap/ColumnarPluginConfig.scala [hunk 1]
```diff
@@ -24,6 +24,8 @@ class ColumnarPluginConfig(conf: SparkConf) {
     conf.getBoolean("spark.sql.columnar.sort", defaultValue = false)
   val enableCodegenHashAggregate: Boolean =
     conf.getBoolean("spark.sql.columnar.codegen.hashAggregate", defaultValue = false)
+  val enableColumnarBroadcastJoin: Boolean =
+    conf.getBoolean("spark.sql.columnar.sort.broadcastJoin", defaultValue = true)
```
Comment (zhouyuan) [inline oap-native-sql/core/src/main/scala/com/intel/oap/ColumnarPluginConfig.scala:28]: let's use unified config names with "oap" inside , maybe spark.oap.columnar.sort.broadcastJoin?

File: oap-native-sql/core/src/main/scala/com/intel/oap/expression/ConverterUtils.scala [hunk 2]
```diff
@@ -53,21 +68,101 @@ object ConverterUtils extends Logging {
...
+    val nodes = new java.util.ArrayList[ArrowFieldNode]()
+    val buffers = new java.util.ArrayList[ArrowBuf]()
+    cols.foreach(vector => {
+      appendNodes(vector.asInstanceOf[FieldVector], nodes, buffers);
     })
-    new ArrowRecordBatch(numRowsInBatch, fieldNodes.toList.asJava, inputData.toList.asJava)
+    new ArrowRecordBatch(numRowsInBatch, nodes, buffers);
+  }
+
+  def appendNodes(
```
Comment (zhouyuan) [inline oap-native-sql/core/src/main/scala/com/intel/oap/expression/ConverterUtils.scala:83]: it looks this piece of code will be used in operators other than BHJ, will this change impact the performance?

File: oap-native-sql/core/src/main/scala/com/intel/oap/expression/ColumnarShuffledHashJoin.scala [hunk 3]
```diff
@@ -177,15 +181,7 @@ object ColumnarShuffledHashJoin extends Logging {
       conditionOption: Option[Expression],
       left: SparkPlan,
       right: SparkPlan,
-      _buildTime: SQLMetric,
```
Comment (zhouyuan) [inline oap-native-sql/core/src/main/scala/com/intel/oap/expression/ColumnarShuffledHashJoin.scala:180]: great cleanup!

File: oap-native-sql/core/src/main/scala/com/intel/oap/execution/ColumnarBroadcastHashJoinExec.scala [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+  val numOutputRows = longMetric("numOutputRows")
+  val totalTime = longMetric("totalTime")
+  val joinTime = longMetric("joinTime")
+  val buildTime = longMetric("buildTime")
+  val fetchTime = longMetric("fetchTime")
+  val resultSchema = this.schema
+
+  //TODO() Disable code generation
+  //override def supportCodegen: Boolean = false
```
Comment (zhouyuan) [inline oap-native-sql/core/src/main/scala/com/intel/oap/execution/ColumnarBroadcastHashJoinExec.scala:99]: disable codegen?

Comment (xuechendi) [reply oap-native-sql/core/src/main/scala/com/intel/oap/execution/ColumnarBroadcastHashJoinExec.scala:99]: @zhouyuan , yes, so base class wholestagecodegen doProduce function will not be called

File: oap-native-sql/core/src/main/scala/com/intel/oap/expression/ConverterUtils.scala (hunk 2 above)
Comment (xuechendi) [reply oap-native-sql/core/src/main/scala/com/intel/oap/expression/ConverterUtils.scala:83]: should not be, logic is pretty much the same

File: oap-native-sql/core/src/main/scala/com/intel/oap/ColumnarPluginConfig.scala (hunk 1 above)
Comment (xuechendi) [reply oap-native-sql/core/src/main/scala/com/intel/oap/ColumnarPluginConfig.scala:28]: can we do this in another PR to clean all configuration?

**Comment by xuechendi** [comment]
@zhouyuan , passed TPCH queries with ColumnarBroadcastHashJoin + AQE, result checked.

**Pull request merged**